Update 05.model-complexity.md (#192)
jaybee84/ml-in-rd: content/04. box_1_experimental_design.md

#### Box 1: Understanding experimental design of machine learning to inform requirements for data

##### Components of ML experiments

Machine learning algorithms identify patterns that explain or fit a given dataset.

> Every machine learning algorithm (_model_) goes through a _training_ phase where it identifies underlying patterns in a given dataset, and a _testing_ phase where the model applies the identified patterns to unseen data points.

Reviewer: ```suggestion
Every machine learning algorithm goes through a _training_ phase where it identifies underlying patterns in a given dataset to create a "trained" algorithm (a _model_) , and a _testing_ phase where the model applies the identified patterns to unseen data points.
```

I'd consider the "model" to be the trained predictor, not the naive algorithm, what do you all think?

> So every model should be provided with the following fundamental parts as input: 1. a _training dataset_ , 2. an _evaluation dataset_ , 3. a _held-out validation dataset_ .

Reviewer: ```suggestion
Typically, a machine learning algorithm is provided with the following fundamental parts as input: 1. a _training dataset_ , 2. an _evaluation dataset_ , 3. a _held-out validation dataset_ .
```

> Typically each of these input datasets consists of a matrix (or dataframe or table) where samples form the rows and features form the columns.

Reviewer: ```suggestion
These input data can be images, text, numbers, or other types of data which are typically encoded as a numerical representation of the input data. 
```

I think this is safe to say is typically true, but there are probably exceptions....

A _training dataset_ is used to expose the model to underlying patterns among the features present in the data of interest.
An _evaluation dataset_ is a small test dataset which is used during the training phase to help the model iteratively update its parameters (i.e., _hyperparameter_tuning_ or _model tuning_).

> In many cases, a large training set may be subdivided to form a smaller training dataset and the evaluation dataset and both used to train the model (see next section for more details on cross-validation).

Reviewer: ```suggestion
In many cases, a large training set may be subdivided to form a smaller training dataset and the evaluation dataset and both used to train the model (see next section for more details on cross-validation).
```
I think this sentence is a bit confusing but not sure how to correct it.

Author: ```suggestion
In many cases, a large training set may be subdivided to form a smaller training dataset and the evaluation dataset, both of which are used to train the model (see next section for more details on cross-validation).
```

> In the testing phase, a new or unseen test dataset or _held-out validation set_ is used to test whether the patterns learned by the model can be generalized to new data points.

Reviewer: ```suggestion
In the testing phase, a new or unseen test dataset or _held-out validation set_ is used to test whether the patterns learned by the model hold true in new data (are _generalizable_).
```

While the evaluation dataset helps the model iteratively update its parameters to learn important patterns in the training data, the held-out validation set helps test the generalizability of the model.

> Generalizability of a model is its ability to recognize patterns that can help predict the class or an outcome for previously unseen data.

Reviewer: This and the next few lines feels like it should be at the end, since this is chronologically at the end of the train-test cycle

This makes sure that all the datasets involved in training and testing a model maintain uniformity in the features.
In case of rare diseases where multiple datasets may be combined to make a large enough training dataset, special care is taken to standardize the features and the patterns therein.
The iterative training stage helps the model learn important patterns in the training dataset and then use the evaluation dataset to test for errors in prediction and update its learning parameters (_hyperparameter tuning_).
The method by which the evaluation dataset tests the performance of the trained model and helps update the hyperparameters is called _cross-validation_.
To maximally utilize the available data for cross-validation, there can be multiple approaches to form the training and evaluation datasets e.g. leave-p-out cross-validation, leave-one-out cross-validation, k-fold cross-validation, Monte-Carlo random subsampling cross-validation [@doi:10.1007/978-1-4614-6849-3].

> In case of k-fold cross-validation, a given dataset is shuffled randomly and split into _k_ parts.

Reviewer: Do we call this out because it's in the figure? or some other reason

Author: not quite sure what you are referring to here.. k-fold cross validation?

Author: we used k-fold cross validation as an example in the text (since it is the easiest to explain :).. and then added this to the figure for illustrative purposes
